releasetasks/util.py

```diff
@@ -66,3 +69,74 @@ def get_json_rev(repo_path, revision):
     req = requests.get(url, timeout=20)
     req.raise_for_status()
     return req.json()
+
+
+def graph_to_tasks(graph):
+    """Convert a graph into a list of tasks.
+    Replaces graph level "requires" into task-level "dependencies"
+    """
+    tasks = OrderedDict()
+    for t in graph["tasks"]:
+        # If a task doesn't have any dependencies it will be scheduled even if
+        # the whole submission fails. To make the submission atomic, we create
+        # a dummy task, which is resolved at the end of the submission to
+        # unblock the scheduling
+        task = t["task"]
+        if t.get("requires"):
+            task["dependencies"] = t["requires"]
+        if task["provisionerId"] == "aws-provisioner-v1":
+            task["payload"]["onExitStatus"] = task["payload"].get("onExitStatus") or \
+                {'retry': [1]}
+
+        tasks[t["taskId"]] = task
+
+    return tasks
+
+
+def add_atomic_task(tasks, toplevel_task):
+    tasks = deepcopy(tasks)
+    tl_task_id, toplevel_task_def = toplevel_task
+    for t in tasks.itervalues():
+        if "dependencies" not in t:
+            t["dependencies"] = [tl_task_id]
+    tasks[tl_task_id] = toplevel_task_def
+    return tasks
+
+
+def sort_tasks(tasks):
+    to_sort = {task_id: set(task.get("dependencies", []))
+               for task_id, task in tasks.iteritems()}
+    ordered = toposort_flatten(to_sort)
+    return OrderedDict(
+        sorted(tasks.items(), key=lambda t: ordered.index(t[0]))
+    )
+
+
+def inject_dummy_tasks(tasks, dummy_task, max_deps=100):
+    new_tasks = OrderedDict()
+    for task_id, task in tasks.iteritems():
+        task = deepcopy(task)
+        deps = task.get("dependencies", [])
+        if len(deps) > max_deps:
+            num_dummy_tasks = int(round(len(deps)/float(max_deps) + 0.5))
```

Comment:
small nit: might be useful to add one-liner comment like "e.g. For a task with 145 dependencies, we need two dummy task". Or just add the top-level comment in to the function explaining why we need these dummy tasks.